src/sensors/msr_rapl.rs

```diff
@@ -0,0 +1,310 @@
+        let power: u32 = new_data & 0x0F;
+
+        //// Intel documentation says: 1 / 2^power
+        let divider = i64::pow(2, power);
+
+        1.0 / divider as f64
+    }
+    pub fn extract_rapl_energy_unit(data: u64) -> f64 {
+        // Intel documentation says high level bits are reserved, so ignore them
+        let new_data: u32 = (data & 0xFFFFFFFF) as u32;
+        //// Energy state units are located from bits 8 to 12, extract them
+        let energy: u32 = (new_data >> 8) & 0x1F;
+
+        //// Intel documentation says: 1 / 2^power
+        let divider = i64::pow(2, energy);
+
+        1.0 / divider as f64
+    }
+    pub fn extract_rapl_time_unit(data: u64) -> f64 {
+        // Intel documentation says high level bits are reserved, so ignore them
+        let new_data: u32 = (data & 0xFFFFFFFF) as u32;
+        //// Time units are located from bits 16 to 19, extract them
+        let time: u32 = (new_data >> 16) & 0x0F;
+
+        //// Intel documentation says: 1 / 2^power
+        let divider = i64::pow(2, time);
+
+        1.0 / divider as f64
+    }
+
+    pub fn extract_rapl_current_power(data: u64, energy_unit: f64) -> String {
+        let energy_consumed: f64 = ((data & 0xFFFFFFFF) as f64) * energy_unit * 1000000.0;
+        format!("{}", energy_consumed as u64)
+    }
+}
+
+impl RecordReader for Topology {
+    fn read_record(&self) -> Result<Record, Box<dyn Error>> {
+        let randval: i32 = rand::random();
+        Ok(Record {
+            timestamp: current_system_time_since_epoch(),
+            unit: super::units::Unit::MicroJoule,
+            value: format!("{}", randval),
+        })
+    }
+}
+
+unsafe fn send_request(
+    device: HANDLE,
+    request_code: u16,
+    request: *const u64,
+    request_length: usize,
+    reply: *mut u64,
+    reply_length: usize,
+) -> Result<String, String> {
+    let mut len: u32 = 0;
+    let len_ptr: *mut u32 = &mut len;
+
+    if DeviceIoControl(
+        device, // envoi 8 octet et je recoi 8 octet
+        crate::sensors::msr_rapl::ctl_code(
+            FILE_DEVICE_UNKNOWN,
+            request_code as _,
+            METHOD_BUFFERED,
+            FILE_READ_DATA.0 | FILE_WRITE_DATA.0, // nouvelle version : METHOD_OUD_DIRECT devien METHOD_BUFFERED
+        ),
+        request as _,
+        request_length as u32,
+        reply as _,
+        reply_length as u32,
+        len_ptr,
+        std::ptr::null_mut(),
+    )
+    .as_bool()
+    {
+        if len != reply_length as u32 {
+            error!(
+                "Got invalid answer length, Expected {}, got {}",
+                reply_length, len
+            );
+        }
+        info!("Device answered");
+        Ok(String::from("Device answered !"))
+    } else {
+        error!("DeviceIoControl failed");
+        Err(String::from("DeviceIoControl failed"))
+    }
+}
+impl RecordReader for CPUSocket {
+    fn read_record(&self) -> Result<Record, Box<dyn Error>> {
```

Comment:
Fonction really hard to read due to nested blocks and formating.
I think a refactor with functions to flatten it would help.

Reply:
It is a bit dummy-written, as I had a hard time at first to make send_request work properly.

I intend to push windows support in 0.5 as an experimental feature, then work on packaging and improvements for the 0.6.

Do you think it could wait the 0.6 ?